File: app/src/main/java/com/vkpapps/soundbooster/fragments/DashboardFragment.java
package com.vkpapps.soundbooster.fragments;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.vkpapps.soundbooster.R;
import com.vkpapps.soundbooster.adapter.ClientAdapter;
import com.vkpapps.soundbooster.connection.ClientHelper;
import com.vkpapps.soundbooster.interfaces.OnFragmentAttachStatusListener;
import com.vkpapps.soundbooster.interfaces.OnNavigationVisibilityListener;
import com.vkpapps.soundbooster.interfaces.OnUserListRequestListener;
import com.vkpapps.soundbooster.interfaces.OnUsersUpdateListener;

import java.util.List;

/**
 * @author VIJAY PATIDAR
 * */
public class DashboardFragment extends Fragment implements OnUsersUpdateListener {

    private List<ClientHelper> users;
    private OnUserListRequestListener onUserListRequestListener;
    private ClientAdapter clientAdapter;
    private OnNavigationVisibilityListener onNavigationVisibilityListener;
    private OnFragmentAttachStatusListener onFragmentAttachStatusListener;

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        //Nothing to display when user is client
        if (users == null) return;

        clientAdapter = new ClientAdapter(users, view.getContext());
        RecyclerView recyclerView = view.findViewById(R.id.clientList);
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        recyclerView.setOnFlingListener(new RecyclerView.OnFlingListener() {
            @Override
            public boolean onFling(int velocityX, int velocityY) {
                if (onNavigationVisibilityListener != null)
                    onNavigationVisibilityListener.onNavVisibilityChange(velocityY < 0);
                return false;
            }
        });
        recyclerView.setAdapter(clientAdapter);
        clientAdapter.notifyDataSetChanged();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_dashboard, container, false);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        onNavigationVisibilityListener = (OnNavigationVisibilityListener) context;
        onUserListRequestListener = (OnUserListRequestListener) context;
        onFragmentAttachStatusListener = (OnFragmentAttachStatusListener) context;
        onFragmentAttachStatusListener.onFragmentAttached(this);
        users = onUserListRequestListener.onRequestUsers();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        onFragmentAttachStatusListener.onFragmentDetached(this);
        onUserListRequestListener = null;
        onNavigationVisibilityListener = null;
        onFragmentAttachStatusListener = null;
    }


    @Override
    public void onUserUpdated() {
        if (clientAdapter != null)
            clientAdapter.notifyDataSetChanged();
    }
}

File: app/src/main/java/com/vkpapps/soundbooster/ui/fragments/DashboardFragment.kt
package com.vkpapps.soundbooster.ui.fragments

import android.content.Context
import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.OnFlingListener
import com.vkpapps.soundbooster.R
import com.vkpapps.soundbooster.connection.ClientHelper
import com.vkpapps.soundbooster.interfaces.OnFragmentAttachStatusListener
import com.vkpapps.soundbooster.interfaces.OnNavigationVisibilityListener
import com.vkpapps.soundbooster.interfaces.OnUserListRequestListener
import com.vkpapps.soundbooster.interfaces.OnUsersUpdateListener
import com.vkpapps.soundbooster.ui.adapter.ClientAdapter

/**
 * @author VIJAY PATIDAR
 */
class DashboardFragment : Fragment(), OnUsersUpdateListener {
    private var users: List<ClientHelper>? = null
    private var onUserListRequestListener: OnUserListRequestListener? = null
    private var clientAdapter: ClientAdapter? = null
    private var onNavigationVisibilityListener: OnNavigationVisibilityListener? = null
    private var onFragmentAttachStatusListener: OnFragmentAttachStatusListener? = null
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_dashboard, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setHasOptionsMenu(true)

        //Nothing to display when user is client
        if (users == null) return
        clientAdapter = ClientAdapter(users, view.context)
        val recyclerView: RecyclerView = view.findViewById(R.id.clientList)
        recyclerView.itemAnimator = DefaultItemAnimator()
        recyclerView.layoutManager = LinearLayoutManager(view.context)
        recyclerView.onFlingListener = object : OnFlingListener() {
            override fun onFling(velocityX: Int, velocityY: Int): Boolean {
                onNavigationVisibilityListener?.onNavVisibilityChange(velocityY < 0)
                return false
            }
        }
        recyclerView.adapter = clientAdapter
        clientAdapter?.notifyDataSetChanged()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.home_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        onNavigationVisibilityListener = context as OnNavigationVisibilityListener
        onUserListRequestListener = context as OnUserListRequestListener
        onFragmentAttachStatusListener = context as OnFragmentAttachStatusListener
        onFragmentAttachStatusListener?.onFragmentAttached(this)
        users = onUserListRequestListener?.onRequestUsers()
    }

    override fun onDetach() {
        super.onDetach()
        onFragmentAttachStatusListener?.onFragmentDetached(this)
        onUserListRequestListener = null
        onNavigationVisibilityListener = null
        onFragmentAttachStatusListener = null
    }

    override fun onUserUpdated() {
        clientAdapter?.notifyDataSetChanged()
    }
}